// amparo-fastify/dist/cjs/fastify/plugins/fastify.plugin.auth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.authPlugin = exports.AuthenticationMethod = void 0;
const fastify_plugin_1 = __importDefault(require("fastify-plugin"));
const core_assert_1 = require("../../core/core.assert");
const cookie_1 = __importDefault(require("@fastify/cookie"));
const core_1 = require("../../core");
var AuthenticationMethod;
(function (AuthenticationMethod) {
    AuthenticationMethod["cookieJwt"] = "cookieJwt";
    AuthenticationMethod["xApiKey"] = "xApiKey";
    AuthenticationMethod["emailPassword"] = "emailPassword";
})(AuthenticationMethod || (exports.AuthenticationMethod = AuthenticationMethod = {}));
const authPlugin = (0, fastify_plugin_1.default)(async (fastify, config) => {
    fastify.decorateRequest("_user");
    fastify.register(cookie_1.default);
    const authenticate = async (request) => {
        const { authenticationMethod, authenticationMethodValue } = (0, core_assert_1.assert)((() => {
            const apiKey = Array.isArray(request.headers["x-api-key"])
                ? request.headers["x-api-key"][0]
                : request.headers["x-api-key"];
            if (apiKey) {
                return {
                    authenticationMethod: "xApiKey" /* AuthenticationMethod.xApiKey */,
                    authenticationMethodValue: apiKey,
                };
            }
            const jwtCookie = request.cookies[config.getUserByAuthMethodHelperOptions?.jwtCookieName ?? "jwt"];
            if (jwtCookie) {
                return {
                    authenticationMethod: "cookieJwt" /* AuthenticationMethod.cookieJwt */,
                    authenticationMethodValue: jwtCookie,
                };
            }
            if (request.body[config.getUserByAuthMethodHelperOptions?.passwordEmailFieldNames
                ?.emailFieldName ?? "email"] &&
                request.body[config.getUserByAuthMethodHelperOptions?.passwordEmailFieldNames
                    ?.passwordFieldName ?? "password"]) {
                return {
                    authenticationMethod: "emailPassword" /* AuthenticationMethod.emailPassword */,
                    authenticationMethodValue: {
                        [config.getUserByAuthMethodHelperOptions
                            ?.passwordEmailFieldNames?.emailFieldName ?? "email"]: request.body[config.getUserByAuthMethodHelperOptions
                            ?.passwordEmailFieldNames?.emailFieldName ?? "email"],
                        [config.getUserByAuthMethodHelperOptions
                            ?.passwordEmailFieldNames?.passwordFieldName ?? "password"]: request.body[config.getUserByAuthMethodHelperOptions
                            ?.passwordEmailFieldNames?.passwordFieldName ?? "password"],
                    },
                };
            }
        })(), "No authentication method found", "authenticationerror" /* ErrorName.authentication */);
        const requestUser = await (0, core_1.safe)(config.getRequestUserByAuthMethodHelper(authenticationMethod, authenticationMethodValue, config.getUserByAuthMethodHelperOptions), "Falid to get user by auth method", "authenticationerror" /* ErrorName.authentication */);
        request._user = requestUser;
    };
    fastify.addHook("onRoute", (routeOptions) => {
        if (routeOptions.config?.authenticate) {
            const preHandler = routeOptions.preHandler;
            if (Array.isArray(preHandler)) {
                routeOptions.preHandler = [authenticate, ...preHandler];
            }
            else if (preHandler) {
                routeOptions.preHandler = [authenticate, preHandler];
            }
            else {
                routeOptions.preHandler = authenticate;
            }
        }
    });
});
exports.authPlugin = authPlugin;
